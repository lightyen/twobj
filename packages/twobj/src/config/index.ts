export { resolveConfig } from "./resolveConfig"
import config from "./defaultConfig"
import colors from "./defaultColors"

const defaultConfig = config
const defaultColors = colors

export { defaultConfig, defaultColors }
