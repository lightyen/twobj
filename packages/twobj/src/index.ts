export { resolveConfig } from "./config/resolveConfig"
export { createContext } from "./core"
export type { CSSProperties, CSSValue } from "./types"
export type { ValueType } from "./values"
import type { CSSProperties } from "./types"

// Fake
export const globalStyles: Record<string, CSSProperties> = {}
export function tw(arr: TemplateStringsArray): CSSProperties {
	return {}
}
export function theme(arr: TemplateStringsArray): unknown {
	return undefined
}
