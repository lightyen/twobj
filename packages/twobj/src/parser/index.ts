export * from "./angle"
export * from "./css"
export * from "./length"
export * from "./nodes"
export * from "./number"
export * from "./parser"
export * from "./percentage"
export * from "./shadow"
export * from "./theme"
export * from "./util"
