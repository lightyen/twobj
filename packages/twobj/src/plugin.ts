/* eslint-disable @typescript-eslint/no-explicit-any */
import type { ConfigJS, CorePlugin, UnnamedPlugin, UserPluginFunctionWithOption, UserPluginObject } from "./types"

interface PluginFunction<Options> {
	(options: Options): UnnamedPlugin
}
interface ConfigFunction<Options> {
	(options: Options): ConfigJS
}

export interface CreatePlugin {
	(handler: UnnamedPlugin): CorePlugin
	(pluginName: string, handler: UnnamedPlugin): CorePlugin
	(handler: UnnamedPlugin, config: ConfigJS): CorePlugin
	withOptions: CreatePluginWithOptions
}

interface CreatePluginWithOptions {
	<Options = unknown>(
		pluginFunction: PluginFunction<Options>,
		configFunction?: ConfigFunction<Options>,
	): UserPluginFunctionWithOption<Options>
	<Options = unknown>(
		pluginName: string,
		pluginFunction: PluginFunction<Options>,
		configFunction?: ConfigFunction<Options>,
	): UserPluginFunctionWithOption<Options>
}

/** Create a tailwind plugin. */
export const plugin: CreatePlugin = (first: string | UnnamedPlugin, second?: unknown, third?: unknown) => {
	// prevent plugin name to be minified
	if (typeof first === "string") {
		Object.defineProperty(second, "name", { value: first, writable: false })
		return second as CorePlugin
	}
	return first
}

plugin.withOptions = function <Options = unknown>(
	first: string | ((options: Options) => UnnamedPlugin),
	second: unknown,
	third?: unknown,
): UserPluginFunctionWithOption<Options> {
	const pluginName = typeof first === "string" ? first : ""

	let pluginFunction: PluginFunction<Options>
	let configFunction: ConfigFunction<Options>
	if (typeof first === "string") {
		pluginFunction = second as PluginFunction<Options>
		configFunction = (third ?? (() => ({}))) as ConfigFunction<Options>
	} else {
		pluginFunction = first
		configFunction = (second ?? (() => ({}))) as ConfigFunction<Options>
	}

	function optionsFunction(options: Options): UserPluginObject {
		return {
			name: pluginName,
			handler: pluginFunction(options),
			config: configFunction(options),
		}
	}

	optionsFunction.__pluginFunction = pluginFunction
	optionsFunction.__configFunction = configFunction
	optionsFunction.__isOptionsFunction = true
	return optionsFunction
}
