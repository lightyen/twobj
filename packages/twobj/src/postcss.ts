import cssesc from "cssesc"
import * as parsel from "parsel-js"
import { escapeRegexp, splitAtTopLevelOnly } from "./parser"

// Postcss source: https://github.com/postcss/postcss-selector-parser/blob/master/src/util/unesc.js

export function escapeCss(value: string): string {
	return cssesc(value, { isIdentifier: true })
}

function gobbleHex(str: string): [string, number] | undefined {
	const lower = str.toLowerCase()
	let hex = ""
	let spaceTerminated = false
	for (let i = 0; i < 6 && lower[i] !== undefined; i++) {
		const code = lower.charCodeAt(i)
		// check to see if we are dealing with a valid hex char [a-f|0-9]
		const valid = (code >= 97 && code <= 102) || (code >= 48 && code <= 57)
		// https://drafts.csswg.org/css-syntax/#consume-escaped-code-point
		spaceTerminated = code === 32
		if (!valid) {
			break
		}
		hex += lower[i]
	}

	if (hex.length === 0) {
		return undefined
	}
	const codePoint = parseInt(hex, 16)

	const isSurrogate = codePoint >= 0xd800 && codePoint <= 0xdfff
	// Add special case for
	// "If this number is zero, or is for a surrogate, or is greater than the maximum allowed code point"
	// https://drafts.csswg.org/css-syntax/#maximum-allowed-code-point
	if (isSurrogate || codePoint === 0x0000 || codePoint > 0x10ffff) {
		return ["\uFFFD", hex.length + (spaceTerminated ? 1 : 0)]
	}

	return [String.fromCodePoint(codePoint), hex.length + (spaceTerminated ? 1 : 0)]
}

const CONTAINS_ESCAPE = /\\/
export function unescapeCss(str: string) {
	const needToProcess = CONTAINS_ESCAPE.test(str)
	if (!needToProcess) {
		return str
	}
	let ret = ""

	for (let i = 0; i < str.length; i++) {
		if (str[i] === "\\") {
			const gobbled = gobbleHex(str.slice(i + 1, i + 7))
			if (gobbled !== undefined) {
				ret += gobbled[0]
				i += gobbled[1]
				continue
			}

			// Retain a pair of \\ if double escaped `\\\\`
			// https://github.com/postcss/postcss-selector-parser/commit/268c9a7656fb53f543dc620aa5b73a30ec3ff20e
			if (str[i + 1] === "\\") {
				ret += "\\"
				i++
				continue
			}

			// if \\ is at the end of the string retain it
			// https://github.com/postcss/postcss-selector-parser/commit/01a6b346e3612ce1ab20219acc26abdc259ccefb
			if (str.length === i + 1) {
				ret += str[i]
			}
			continue
		}

		ret += str[i]
	}

	return ret
}

export function findClasses(selector: string): Map<string, string> {
	interface OtherNode {
		type: string
		list?: Node[]
		subtree?: Node
		left?: Node
		right?: Node
	}

	interface ClassNode {
		type: "class"
		content: string
		name: string
		pos: [number, number]
	}

	type Node = ClassNode & OtherNode

	const classes = new Map<string, string>()

	splitAtTopLevelOnly(selector).forEach(s => walk(parsel.parse(s.value)))

	return classes

	function isExist(node: parsel.AST | null | undefined): node is parsel.AST {
		return typeof node === "object" && node != null
	}

	function isTokens(node: parsel.AST): node is parsel.Tokens {
		switch (node.type) {
			case "complex":
			case "compound":
			case "list":
				return false
			default:
				return true
		}
	}

	function isList(node: parsel.AST): node is parsel.List {
		return node.type === "list" || Array.isArray(node["list"])
	}

	function isComplex(node: parsel.AST): node is parsel.Complex {
		return node.type === "complex" || node["left"] != null || node["right"] != null
	}

	function walk(node: parsel.AST | null | undefined) {
		if (!isExist(node)) {
			return
		}

		if (callback(node) === false) {
			return
		}

		if (isList(node)) {
			node.list.forEach(walk)
			return
		}

		if (isTokens(node)) {
			walk(node.subtree)
			return
		}

		if (isComplex(node) && isExist(node.left)) {
			walk(node.left)
		}

		if (isComplex(node) && isExist(node.right)) {
			walk(node.right)
		}
	}

	function callback(node: parsel.AST): boolean | void {
		if (node.type === "class") {
			if (!classes.has(node.name)) {
				const replaced = selector.replace(new RegExp(`[.]${escapeRegexp(node.name)}(?!-)\\b`, "g"), "&")
				classes.set(unescapeCss(node.name), replaced)
			}
			return false
		}
	}
}
