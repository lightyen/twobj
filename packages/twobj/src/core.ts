import { classPlugins } from "./classPlugins"
import { camelCase, createParser, findRightBracket } from "./parser"
import {
	ArbitraryClassname,
	ArbitrarySelector,
	ArbitraryVariant,
	Classname,
	Expression,
	NodeType,
	SimpleVariant,
	Variant,
} from "./parser/nodes"
import * as theme from "./parser/theme"
import { isPluginWithOptions } from "./plugin"
import { escapeCss, findClasses } from "./postcss"
import type {
	ConfigObject,
	Context,
	CorePluginFeatures,
	CorePluginOptions,
	CSSProperties,
	CSSValue,
	LookupSpec,
	Palette,
	PostModifier,
	ResolvedConfigJS,
	StaticSpec,
	UserPluginOptions,
	ValueType,
	VariantSpec,
} from "./types"
import {
	applyCamelCase,
	applyImportant,
	applyModifier,
	flattenColorPalette,
	getAmbiguousFrom,
	getClassListFrom,
	getColorClassesFrom,
	isCSSValue,
	isFunction,
	isNotEmpty,
	isObject,
	isString,
	merge,
	toArray,
} from "./util"
import { representAny, representTypes } from "./values"
import { variantPlugins } from "./variantPlugins"

/** Create a tailwind context. */
export function createContext(config: ResolvedConfigJS): Context {
	if (typeof config.separator !== "string" || !config.separator) {
		config.separator = ":"
	}

	const parser = createParser()
	parser.separator = config.separator

	if (typeof config.prefix !== "string") {
		config.prefix = ""
	}

	const importantAll = config.important === true

	const corePlugins = new Set(
		(() => {
			const { corePlugins } = config
			if (Array.isArray(corePlugins)) {
				return corePlugins
			}
			if (isObject(corePlugins)) {
				return Object.values(classPlugins)
					.map(p => p.name)
					.filter(Boolean)
					.filter(k => corePlugins[k] !== false)
			}
			if (typeof corePlugins === "boolean" && corePlugins === false) {
				return []
			}
			return Object.values(classPlugins).map(p => p.name)
		})(),
	)

	const globalStyles: Record<string, CSSProperties> = {
		"*, ::before, ::after": {},
		"::backdrop": {},
	}

	const defaults = [globalStyles["*, ::before, ::after"], globalStyles["::backdrop"]]
	const utilityMap = new Map<string, LookupSpec | StaticSpec | Array<LookupSpec | StaticSpec>>()
	const variantMap = new Map<string, VariantSpec>()
	const arbitraryVariantMap = new Map<string, (value: string) => VariantSpec>()
	const arbitraryUtilityMap = new Map<string, Set<ValueType | "any">>()

	const options: UserPluginOptions = {
		addBase,
		addDefaults,
		addUtilities,
		addComponents,
		addVariant,
		matchUtilities,
		matchComponents,
		matchVariant,
		config: legacyConfig,
		theme: resolveTheme,
		e(value) {
			return escapeCss(value)
		},
		variants(corePlugin) {
			return []
		},
		corePlugins(feature: keyof CorePluginFeatures): boolean {
			return corePlugins.has(feature)
		},
		prefix(value) {
			return value
		},
	}

	// resolvePlugins

	const apiContext: CorePluginOptions = {
		...options,
		configObject: config,
		themeObject: config.theme,
	}

	const features = new Set<string>()
	let currentPluginName: string | undefined

	for (const [, plugin] of Object.entries(classPlugins)) {
		currentPluginName = plugin.name
		if (corePlugins.has(currentPluginName)) {
			features.add(currentPluginName)
			plugin(apiContext)
		}
		currentPluginName = undefined
	}

	for (const [, plugin] of Object.entries(variantPlugins)) {
		plugin(apiContext)
	}

	for (let i = 0; i < config.plugins.length; i++) {
		let plugin = config.plugins[i]
		if (isPluginWithOptions(plugin)) {
			plugin = plugin()
		}

		const pluginName = plugin["name"]
		if (isString(pluginName) && isNotEmpty(pluginName)) {
			currentPluginName = pluginName
		} else {
			currentPluginName = `userPlugin-${i}`
		}
		features.add(currentPluginName)

		if (isFunction(plugin)) {
			plugin(options)
		} else if (plugin.handler) {
			plugin.handler(options)
		}

		currentPluginName = undefined
	}

	resolveGlobalTheme(globalStyles)

	return {
		parser,
		globalStyles,
		utilities: utilityMap,
		variantMap,
		arbitraryVariants: arbitraryVariantMap,
		arbitraryUtilities: arbitraryUtilityMap,
		css,
		getPluginName,
		features,
		renderTheme(value) {
			return theme.renderTheme(config, value)
		},
		renderThemeFunc,
		getClassList() {
			return getClassListFrom(utilityMap)
		},
		getColorClasses() {
			return getColorClassesFrom(utilityMap)
		},
		getAmbiguous() {
			return getAmbiguousFrom(utilityMap)
		},
		cssVariant,
		...options,
	}

	function renderThemeFunc(value: string): string {
		return theme.renderThemeFunc(config, value)
	}
	function resolveTheme(value: string, defaultValue?: unknown): unknown {
		return theme.resolveThemeNoDefault(config, value, defaultValue)
	}

	function resolveGlobalTheme(g: CSSProperties): void {
		if (typeof g !== "object") return
		Object.entries(g).map(([key, child]) => {
			if (typeof child !== "object") {
				if (typeof child === "string") {
					g[key] = theme.renderThemeFunc(config, child)
				}
				return
			}
			resolveGlobalTheme(child)
		})
	}

	function legacyConfig(path: string, defaultValue?: unknown): unknown {
		const node = theme.parse_theme_val(path)
		const useDefault = true
		const target = theme.resolve(config, node.path, useDefault)
		return target === undefined ? defaultValue : target
	}

	function addUtilitySpec(key: string, core: StaticSpec | LookupSpec): void {
		const obj = utilityMap.get(key)
		if (obj == null) {
			utilityMap.set(key, core)
			return
		}
		if (Array.isArray(obj)) {
			obj.push(core)
			return
		}
		utilityMap.set(key, [obj, core])
	}

	function addBase(bases: CSSProperties | CSSProperties[]): void {
		bases = toArray(bases).map(applyCamelCase).map(expandAtRules)
		merge(globalStyles, ...bases)
	}

	function addDefaults(pluginName: string, properties: Record<string, string | string[]>): void {
		properties = Object.fromEntries(Object.entries(properties).map(([key, value]) => [camelCase(key), value]))
		for (let i = 0; i < defaults.length; i++) {
			merge(defaults[i], properties)
		}
	}

	function addVariant(
		variantName: string,
		variantDesc: string | string[],
		options: {
			postModifier?: PostModifier
		} = {},
	): void {
		if (variantMap.has(variantName)) throw Error(`variant '${variantName} is duplicated.'`)
		variantDesc = toArray(variantDesc)
		if (variantDesc.some(value => typeof value !== "string"))
			throw Error(`variant description type should be string or string[].`)
		variantDesc = variantDesc.map(v => v.replace(/:merge\((.*?)\)/g, "$1"))
		variantMap.set(variantName, createVariantSpec(variantDesc, options.postModifier))
	}

	function matchVariant(
		variants: Record<string, (value: string) => string | string[]>,
		options: {
			values?: Record<string, string>
			postModifier?: PostModifier
		} = {},
	): void {
		for (const key in variants) {
			for (const [k, v] of Object.entries(options.values ?? {})) {
				addVariant(`${key}-${k}`, variants[key](v), { postModifier: options.postModifier })
			}
			arbitraryVariantMap.set(key, value =>
				createVariantSpec(toArray(variants[key](value)), options.postModifier),
			)
		}
	}

	function createVariantSpec(variantDesc: string[], postModifier?: PostModifier): VariantSpec {
		const fns = variantDesc.map<VariantSpec>(desc => {
			const reg = /{/gs
			const match = reg.exec(desc)
			if (!match) {
				return (css = {}) => ({ [desc]: typeof css === "object" ? { ...css } : css })
			} else {
				const rb = findRightBracket({ text: desc, start: reg.lastIndex - 1, brackets: [123, 125] })
				if (rb == undefined) {
					return (css = {}) => ({ [desc]: css })
				}
				const scope = desc
					.slice(0, reg.lastIndex - 1)
					.trim()
					.replace(/\s{2,}/g, " ")
				const restDesc = desc.slice(reg.lastIndex, rb).trim()
				if (scope) {
					return (css = {}) => ({ [scope]: createVariantSpec([restDesc])(css) })
				} else {
					return (css = {}) => createVariantSpec([restDesc])(css)
				}
			}
		})
		return (css = {}) => {
			if (postModifier) {
				css = applyModifier(css, postModifier)
			}
			return Object.assign({}, ...fns.map(fn => fn(css)))
		}
	}

	function addComponents(
		components: CSSProperties | CSSProperties[],
		{
			respectPrefix = true,
			respectImportant = false,
		}: { respectPrefix?: boolean; respectImportant?: boolean } = {},
	): void {
		addUtilities(components, { respectPrefix, respectImportant })
	}

	function addUtilities(
		utilities: CSSProperties | CSSProperties[],
		{ respectPrefix = true, respectImportant = true }: { respectPrefix?: boolean; respectImportant?: boolean } = {},
	): void {
		utilities = toArray(utilities).map(applyCamelCase)
		const keyStylePairs = getKeyStylePairs(utilities)
		for (const [_key, css] of keyStylePairs) {
			const key = respectPrefix ? config.prefix + _key : _key
			addUtilitySpec(key, {
				type: "static",
				css: expandAtRules(css),
				supportsNegativeValues: false,
				pluginName: currentPluginName,
				respectPrefix,
				respectImportant,
			})
		}
		return
	}

	function getKeyStylePairs(styles: CSSProperties[]): Map<string, CSSProperties> {
		const ret = new Map<string, CSSProperties>()
		const result = styles.flatMap(s => traverse(s)).filter((v): v is [string, CSSProperties] => v != undefined)
		for (const [key, css] of result) {
			if (ret.has(key)) {
				ret.set(key, merge(ret.get(key), css))
			} else {
				ret.set(key, css)
			}
		}

		return ret

		function traverse(css: CSSProperties[string]): Array<[string, CSSProperties]> | undefined {
			if (typeof css !== "object" || css === null) {
				return undefined
			}
			const ret: Array<[string, CSSProperties]> = []

			for (const selector in css) {
				const selectors = findClasses(selector)
				if (selectors.size === 0) {
					const result = traverse(css[selector])
					if (result != undefined) {
						result.map(([key, css]) => {
							ret.push([key, { [selector]: css }])
						})
					}
				} else {
					const value = css[selector]
					for (const [key, rest] of selectors) {
						if (rest !== "&") {
							ret.push([key, { [rest]: value }])
						} else if (!isCSSValue(value)) {
							ret.push([key, value])
						}
					}
				}
			}
			return ret
		}
	}

	function expandAtRules(style: CSSProperties | undefined = {}) {
		// { "& > :not([hidden]) ~ :not([hidden])": { borderColor: "black" } }
		// { "@apply text-white bg-black": { borderColor: "black" } }
		// { "@media (min-width: 1024px)": { ".container": { padding: "10px" } } }
		// { "@media (min-width: 1024px)": { "&": { padding: "10px" } } }
		// { "@screen md": { borderColor: "black" } }
		// { "@screen md": { "@apply text-white bg-black": { borderColor: "black" } } }
		const AT_APPLY = /^@apply\s+/
		const AT_SCREEN = /^@screen\s+/

		const target: CSSProperties = {}
		for (const [key, value] of Object.entries(style)) {
			if (isCSSValue(value)) {
				target[key] = value
				continue
			}

			if (AT_APPLY.test(key)) {
				const input = key.replace(AT_APPLY, "")
				merge(target, merge(css(input), expandAtRules(value)))
				continue
			}

			if (AT_SCREEN.test(key)) {
				const input = key.replace(AT_SCREEN, "")
				const fn = compose(variantMap.get(input), expandAtRules)
				merge(target, fn(value))
				continue
			}

			target[key] = expandAtRules(value)
		}
		return target
	}

	function matchComponents(
		components: Record<string, (value: CSSValue) => CSSProperties | CSSProperties[]>,
		{
			type = "any",
			values = {},
			supportsNegativeValues = false,
			filterDefault = false,
			respectPrefix = true,
			respectImportant = false,
		}: {
			type?: (ValueType | "any") | (ValueType | "any")[]
			values?: ConfigObject
			supportsNegativeValues?: boolean
			filterDefault?: boolean
			respectPrefix?: boolean
			respectImportant?: boolean
		} = {},
	): void {
		matchUtilities(components, {
			type,
			values,
			supportsNegativeValues,
			filterDefault,
			respectPrefix,
			respectImportant,
		})
	}

	function matchUtilities(
		utilities: Record<string, (value: CSSValue) => CSSProperties | CSSProperties[]>,
		{
			type = "any",
			values = {},
			supportsNegativeValues = false,
			filterDefault = false,
			respectPrefix = true,
			respectImportant = true,
		}: {
			type?: (ValueType | "any") | (ValueType | "any")[]
			values?: ConfigObject
			supportsNegativeValues?: boolean
			filterDefault?: boolean
			respectPrefix?: boolean
			respectImportant?: boolean
		} = {},
	): void {
		for (const [_key, fn] of Object.entries(utilities)) {
			const key = respectPrefix ? config.prefix + _key : _key
			let represent: LookupSpec["represent"]
			const types = toArray(type)
			if (types.some(t => t === "any")) {
				let ambiguous = false
				let spec = utilityMap.get(key)
				if (spec) {
					spec = toArray(spec).filter((s): s is LookupSpec => s.type === "lookup")
					ambiguous = spec.length > 1
				}

				represent = (input, node, negative) => {
					if (negative && !supportsNegativeValues) return undefined
					return representAny({
						input,
						node,
						values,
						negative,
						ambiguous,
						template(value) {
							const css = fn(value)
							return merge({}, ...toArray(css).map(applyCamelCase))
						},
						filterDefault,
					})
				}
				addUtilitySpec(key, {
					type: "lookup",
					values: values ?? {},
					represent,
					supportsNegativeValues,
					filterDefault,
					pluginName: currentPluginName,
					respectPrefix,
					respectImportant,
				})
				const valueTypes = arbitraryUtilityMap.get(key)
				if (valueTypes) {
					valueTypes.add("any")
				} else {
					arbitraryUtilityMap.set(key, new Set(["any"]))
				}
				continue
			}

			// if types has 'color', always flatten the values
			let isColor = false
			if (types.some(t => t === "color")) {
				values = flattenColorPalette(values as Palette)
				isColor = true
			}

			const noAnyTypes = types.filter((t): t is ValueType => t !== "any")
			represent = (input, node, negative) => {
				if (negative && !supportsNegativeValues) return undefined

				let ambiguous = false
				let spec = utilityMap.get(key)
				if (spec) {
					spec = toArray(spec).filter((s): s is LookupSpec => s.type === "lookup")
					ambiguous = spec.length > 1
				}

				return representTypes({
					input,
					node,
					values,
					negative,
					types: noAnyTypes,
					filterDefault,
					ambiguous,
					template(value) {
						const css = fn(value)
						return merge({}, ...toArray(css))
					},
				})
			}
			addUtilitySpec(key, {
				type: "lookup",
				values: values ?? {},
				represent,
				supportsNegativeValues,
				filterDefault,
				isColor,
				pluginName: currentPluginName,
				respectPrefix,
				respectImportant,
			})
			const valueTypes = arbitraryUtilityMap.get(key)
			if (valueTypes) {
				noAnyTypes.forEach(type => {
					valueTypes.add(type)
				})
			} else {
				arbitraryUtilityMap.set(key, new Set(noAnyTypes))
			}
		}
	}

	function compose(...variants: (VariantSpec | undefined)[]): VariantSpec {
		let wrap: VariantSpec = (css = {}) => css
		for (const f of variants.reverse()) {
			if (f) {
				const g = wrap
				wrap = (css = {}) => f(g(css))
			}
		}
		return wrap
	}

	function getPluginName(value: string): string | undefined {
		const program = parser.parse(value)
		if (program.expressions.length !== 1) {
			return undefined
		}

		const node = program.expressions[0]
		if (node.type !== NodeType.ClassName && node.type !== NodeType.ArbitraryClassname) {
			return undefined
		}

		const [, spec] = classname(node)
		return spec?.pluginName
	}

	function css(strings: string): CSSProperties
	function css(strings: TemplateStringsArray): CSSProperties
	function css(strings: string | TemplateStringsArray): CSSProperties {
		let value = ""
		if (typeof strings !== "string") {
			value = strings[0] as string
		} else {
			value = strings
		}
		const rootStyle: CSSProperties = {}
		let importantRootStyle: CSSProperties | undefined

		const importantSelector: string = typeof config.important === "string" ? config.important : ""
		if (importantSelector) {
			importantRootStyle = {}
			rootStyle[`${importantSelector} &`] = importantRootStyle
		}

		const rootFn: VariantSpec = (css = {}) => css
		const program = parser.parse(value)
		for (let i = 0; i < program.expressions.length; i++) {
			process(program.expressions[i], rootStyle, importantRootStyle, rootFn, false)
		}
		return rootStyle
	}

	function process(
		node: Expression,
		root: CSSProperties,
		importantRoot: CSSProperties | undefined,
		variantCtx: VariantSpec,
		important: boolean,
	) {
		switch (node.type) {
			case NodeType.VariantSpan: {
				switch (node.variant.type) {
					case NodeType.SimpleVariant: {
						const variant = simpleVariant(node.variant)
						if (node.child) {
							process(node.child, root, importantRoot, compose(variantCtx, variant), important)
						}
						break
					}
					case NodeType.ArbitraryVariant: {
						const variant = arbitraryVariant(node.variant)
						if (node.child) {
							process(node.child, root, importantRoot, compose(variantCtx, variant), important)
						}
						break
					}
					case NodeType.ArbitrarySelector: {
						const variant = arbitrarySelector(node.variant)
						if (node.child) {
							process(node.child, root, importantRoot, compose(variantCtx, variant), important)
						}
						break
					}
				}
				break
			}
			case NodeType.Group: {
				for (let i = 0; i < node.expressions.length; i++) {
					process(node.expressions[i], root, importantRoot, variantCtx, important || node.important)
				}
				break
			}
			case NodeType.ClassName:
			case NodeType.ArbitraryClassname: {
				const [result, spec] = classname(node)
				let css = result
				if (css != undefined) {
					if (important || node.important || (spec?.respectImportant && importantAll)) {
						css = applyImportant(css)
					}
				}
				if (spec?.respectImportant && importantRoot) {
					merge(importantRoot, variantCtx(css))
				} else {
					merge(root, variantCtx(css))
				}
				break
			}
			case NodeType.ArbitraryProperty: {
				const i = node.decl.getText().indexOf(":")
				if (i !== -1) {
					const prop = node.decl.getText().slice(0, i).trim()
					const value = renderThemeFunc(node.decl.getText().slice(i + 1)).trim()
					if (prop && value) {
						let css: CSSProperties = { [camelCase(prop)]: value }

						if (css != undefined) {
							if (important || node.important || importantAll) {
								css = applyImportant(css)
							}
						}
						if (importantRoot) {
							merge(importantRoot, variantCtx(css))
						} else {
							merge(root, variantCtx(css))
						}
					}
				}
				break
			}
			case NodeType.ShortCss: {
				// do nothing
				break
			}
		}
	}

	function simpleVariant(node: SimpleVariant) {
		return variantMap.get(node.id.getText())
	}

	function arbitraryVariant(node: ArbitraryVariant) {
		let variant: VariantSpec | undefined
		const prefix = node.prefix.getText()
		if (prefix.endsWith("-")) {
			const key = prefix.slice(0, -1)
			const spec = arbitraryVariantMap.get(key)
			if (spec) {
				const input = node.selector.getText()
				variant = spec(input)
			}
		}
		return variant
	}

	function arbitrarySelector(node: ArbitrarySelector) {
		const value = node.selector
			.getText()
			.trim()
			.replace(/\s{2,}/g, " ")
		if (value) {
			const variant: VariantSpec = (css = {}) => ({ [value]: css })
			return variant
		}
		return undefined
	}

	function classname(node: Classname | ArbitraryClassname): [css?: CSSProperties, spec?: LookupSpec | StaticSpec] {
		let value: string
		if (node.type === NodeType.ClassName) {
			value = node.getText()
			if (value === "$e") {
				return [Math.E as unknown as CSSProperties]
			}
		} else {
			value = node.prefix.getText()
			if (node.expr) {
				node.expr.value = renderThemeFunc(node.expr.value)
			}
		}
		const { spec, negative, restInput } = parseInput(value)

		if (spec) {
			const arr = toArray(spec)
			if (arr.length === 1) {
				const c = arr[0]
				if (c.type === "lookup") {
					const css = c.represent(restInput, node, negative)
					if (css) {
						return [css, c]
					}
				} else {
					return [c.css, c]
				}
			} else if (arr.filter(c => c.type === "lookup").length > 1) {
				const lookup = arr.filter((c): c is LookupSpec => c.type === "lookup")
				const result = lookup
					.map(c => {
						const css = c.represent(restInput, node, negative)
						if (css) {
							return [css, c]
						}
						return undefined
					})
					.filter((t): t is [CSSProperties, LookupSpec] => !!t)
				if (result.length > 1) {
					return []
				} else if (result.length === 1) {
					return result[0]
				}

				const statik = arr.filter((c): c is StaticSpec => c.type === "static")
				if (statik.length > 0) {
					const c = statik[0]
					return [c.css, c]
				}
			} else {
				for (const c of arr) {
					if (c.type === "lookup") {
						const css = c.represent(restInput, node, negative)
						if (css) {
							return [css, c]
						}
					} else {
						return [c.css, c]
					}
				}
			}
		}
		return []
	}

	function parseInput(input: string) {
		const spec = utilityMap.get(input)
		const result = { spec, negative: false, restInput: "" }

		if (spec) {
			return result
		}

		if (input[0] === "-") {
			result.negative = true
			input = input.slice(1)
			const spec = utilityMap.get(input)
			if (spec) {
				result.spec = spec
				result.restInput = input
				return result
			}
		}

		let i = input.length
		while (i > 0) {
			i = input.lastIndexOf("-", i - 1)
			if (i === -1) {
				break
			}

			const key = input.slice(0, i)
			const spec = utilityMap.get(key)
			if (!spec) {
				continue
			}

			const lookupSpec = toArray(spec).filter((s): s is LookupSpec => s.type === "lookup")
			if (lookupSpec.length === 0) {
				continue
			}

			result.spec = lookupSpec
			result.restInput = input.slice(i + 1)
			return result
		}

		return result
	}

	function cssVariant(...variants: Array<Variant | string>) {
		return compose(
			...variants.map(value => {
				if (typeof value === "string") {
					const variant = variantMap.get(value)
					if (variant) {
						return variant
					}

					const program = parser.parse(value + config.separator)
					if (program.expressions.length !== 1) {
						return undefined
					}

					const node = program.expressions[0]
					if (node.type !== NodeType.VariantSpan || node.variant.type === NodeType.SimpleVariant) {
						return undefined
					}

					if (node.variant.type === NodeType.ArbitrarySelector) {
						return arbitrarySelector(node.variant)
					}

					return arbitraryVariant(node.variant)
				}

				const node = value
				if (node.type === NodeType.SimpleVariant) {
					return variantMap.get(node.id.getText())
				}

				if (node.type === NodeType.ArbitrarySelector) {
					return arbitrarySelector(node)
				}

				return arbitraryVariant(node)
			}),
		)
	}
}
