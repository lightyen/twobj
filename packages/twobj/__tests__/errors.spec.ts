import { createContext, resolveConfig } from "../src"

test("throw errors", async () => {
	const ctx = createContext(resolveConfig({ corePlugins: { textColor: false } }))
	expect(ctx.css("text-black")).toEqual({})
	expect(ctx.css("accent-black")).toEqual({ accentColor: "#000" })
	expect(ctx.css("first:text-black")).toEqual({ "&:first-child": {} })
	expect(ctx.css("first:accent-black")).toEqual({ "&:first-child": { accentColor: "#000" } })

	ctx.throwError = true
	expect(() => {
		ctx.css(" (text-black) ")
	}).toThrowErrorMatchingSnapshot()
	expect(ctx.css("accent-black")).toEqual({ accentColor: "#000" })
	expect(() => {
		ctx.css("  first:(text-black)  ")
	}).toThrowErrorMatchingSnapshot()
	expect(ctx.css("first:accent-black")).toEqual({ "&:first-child": { accentColor: "#000" } })
	expect(() => {
		ctx.css("second:accent-black")
	}).toThrowErrorMatchingSnapshot()
	expect(() => {
		ctx.css("(second:):accent-black")
	}).toThrowErrorMatchingSnapshot()
})
