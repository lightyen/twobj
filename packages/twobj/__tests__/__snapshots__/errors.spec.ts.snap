// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throw errors 1`] = `"Utility is not found."`;

exports[`throw errors 2`] = `"Utility is not found."`;

exports[`throw errors 3`] = `"Variant is not found."`;

exports[`throw errors 4`] = `"Variant is not found."`;

exports[`throw errors 5`] = `"Bracket is not closed."`;

exports[`throw errors 6`] = `"Bracket is not closed."`;

exports[`throw errors 7`] = `"Bracket is not closed."`;

exports[`throw errors 8`] = `"Not supported."`;

exports[`throw errors 9`] = `"Not supported."`;
